### PR TITLE
chore: Remove ChainSpec from PayloadConfig 

### LINKED ISSUE #10951: remove chainspec from PayloadConfig
### Describe the feature

~~blocked by #10950~~
ref #10510

ideally we don't need the chainspec here:

https://github.com/paradigmxyz/reth/blob/abd4642d65b07be08debb67d2b5e9a5b3a403f32/crates/payload/basic/src/lib.rs#L675-L675

if we use eth EvmConfig to fill all the settings

## TODO
* remove once we have #10950

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by garwahl**
Closes https://github.com/paradigmxyz/reth/issues/10951

Removes `ChainSpec` from `PayloadConfig`. Where a ChainSpec is required, we instead enforce a `ChainSpecProvider` trait bound on the Client and grab it from there. Updates usages accordingly.

### CODE REVIEW

File: crates/ethereum/payload/src/lib.rs [hunk 1]
```diff
@@ -314,6 +314,7 @@ where
     }
 
     // calculate the requests and the requests root
+    // TODO (garwah): Figure out how to replace this check now that chain spec has been removed.
```
Comment (garwahl) [inline crates/ethereum/payload/src/lib.rs:317]: @mattsse It looks like `chainspec` is used heavily to determine what hardforks are active here, or threaded down eg. https://github.com/paradigmxyz/reth/blob/afac7e04af1534604bac6c76af71f2e886c605ae/crates/evm/src/system_calls/eip4788.rs#L74-L76

Do we have an alternative here for these checks or could we pull a chainspec from somewhere else?

File: examples/custom-engine-types/src/main.rs [hunk 2]
```diff
@@ -312,9 +312,10 @@ where
         client: &Client,
         config: PayloadConfig<Self::Attributes>,
     ) -> Result<Self::BuiltPayload, PayloadBuilderError> {
-        let PayloadConfig { parent_block, extra_data, attributes, chain_spec } = config;
+        let PayloadConfig { parent_block, extra_data, attributes} = config;
+        // TODO (garwah): How do we instantiate EthEvmConfig as it requires chain spec
         <reth_ethereum_payload_builder::EthereumPayloadBuilder as PayloadBuilder<Pool, Client>>::build_empty_payload(&reth_ethereum_payload_builder::EthereumPayloadBuilder::new(EthEvmConfig::new(chain_spec.clone())),client,
```
Comment (garwahl) [inline examples/custom-engine-types/src/main.rs:317]: Similar issue here, it looks like pulling out `chainspec` will break instantiating a EthEvmConfig, any guidance here?

**Review by garwahl — COMMENTED** [review]
Raised this draft to get some guidance as removing `chainspec` seems to break a few things, keen to understand either where we can pull it from or replace usages with something equiv. cc @mattsse

File: crates/ethereum/payload/src/lib.rs (hunk 1 above)
Comment (garwahl) [reply crates/ethereum/payload/src/lib.rs:317]: Alternatively, we move the `EthereumHardForks` helper methods off ChainSpec onto something else

**Review by mattsse — CHANGES_REQUESTED** [review]
I see, you can work around this by requiring that the client is also `ChainSpecProvider` and then get the chainspec from there instead

https://github.com/paradigmxyz/reth/blob/242a3ab06befbd06a51cad1602ca80df6930a4cb/crates/ethereum/payload/src/lib.rs#L90-L90

which should work because this is a chainspec provider

**Review by mattsse — APPROVED** [review]

**Pull request merged**